### DISCUSSION TITLE
erase data under classname

### DISCUSSION THREAD

**Discussion by IngridJSJ** [Spine-project/Spine-Toolbox#1739] [Discussion]
Hello, 
To make simulations faster, I would like to delete a classname and all the data from a python script, once that data is no longer in use. 
Could you please give me a quick example of how to do that? 
Thank you very much!

**Comment by soininen** [comment]
So, you want to remove a class, its objects, parameters and parameter values from a Spine database, am I right? In that case, the following script might be a good starting point:

```python
from spinedb_api import DatabaseMapping
class_name = "class_to_remove"
class_type = "object_class"  # or "relationship_class"
url = "sqlite:///my_database.sqlite"
db_map = DatabaseMapping(url)


def class_id_for_name(name):
    return db_map.query(db_map.entity_class_sq).filter(db_map.entity_class_sq.c.name == name).first().id


try:
    id_to_remove = class_id_for_name(class_name)
    db_map.cascade_remove_items(**{class_type: {id_to_remove}})
    db_map.commit_session(f"Removed {class_name}")
finally:
    db_map.connection.close()
```

Since data is removed in a cascade, removing the class will also remove all its entities (objects/relationships), parameters and parameter values. Note, that in case you remove an object class, all relationship classes (and their associated data) that have the said object class will get removed as well.

**Comment by IngridJSJ** [comment]
Hello, 
I am trying it, but it gets stuck while quering the id_to_remove with class_id_for_name. 
I am passing
class_name = "Bids"
also, what should be the class_type?
class_type = "object_class"

![image](https://user-images.githubusercontent.com/47012366/188115028-275eb557-7c32-445c-b3f2-ca2512d916d7.png)

For other cases, can I also erase some objects according to the parameter_value? For example, erase this bid is status is not "Accepted". 
Thank you very much!

**Reply by soininen** [reply]
Do you get some Traceback/error message? Or does it just hang there?

The `class_name` and `class_type` you're using look correct.

To erase objects according to parameter value, you need to query parameter values and pick the corresponding object ids. From the top of my head (warning: untested code):

```python
from spinedb_api import DatabaseMapping, from_database

url = "sqlite:///path/to/database.sqlite"
db_map = DatabaseMapping(url)
try:
    subquery = db_map.object_parameter_value_sq
    statuses = {row.object_id: from_database(row.value, row.type) for row in db_map.query(subquery).filter(subquery.c.parameter_name == "status")}
    removable_object_ids = {object_id for object_id, status in statuses.items() if status != "Accepted"}
    db_map.cascade_remove_items(object=removable_object_ids)
    db_map.commit_session("Removed unacceptable objects.")
finally:
    db_map.connection.close()
```

**Reply by IngridJSJ** [reply]
I think it was hanging because I opened th DB in a previous step db_emlab = SpineDB(db_url)
As a standalone function works perfect. 
Thank you very much!
